File: snippets/screens/flutter/followers_screen/1/followers_screen.dart
import 'package:flutter/material.dart';

class FollowersScreen extends StatefulWidget {
  final String username;

  const FollowersScreen({Key? key, required this.username}) : super(key: key);

  @override
  _FollowersScreenState createState() => _FollowersScreenState();
}

class _FollowersScreenState extends State<FollowersScreen> {
  final TextEditingController _searchController = TextEditingController();
  List<User> followers = [
    User(
      id: '1',
      username: 'alice_wonder',
      displayName: 'Alice Wonder',
      avatar: 'https://example.com/avatar3.jpg',
      isFollowing: true,
      isFollowedBy: true,
    ),
    User(
      id: '2',
      username: 'bob_builder',
      displayName: 'Bob Builder',
      avatar: 'https://example.com/avatar4.jpg',
      isFollowing: false,
      isFollowedBy: true,
    ),
    User(
      id: '3',
      username: 'charlie_brown',
      displayName: 'Charlie Brown',
      avatar: 'https://example.com/avatar5.jpg',
      isFollowing: true,
      isFollowedBy: true,
    ),
    User(
      id: '4',
      username: 'diana_prince',
      displayName: 'Diana Prince',
      avatar: 'https://example.com/avatar6.jpg',
      isFollowing: false,
      isFollowedBy: true,
    ),
    User(
      id: '5',
      username: 'edward_stark',
      displayName: 'Edward Stark',
      avatar: 'https://example.com/avatar7.jpg',
      isFollowing: true,
      isFollowedBy: true,
    ),
  ];

  List<User> filteredFollowers = [];
  bool isLoading = false;

  @override
  void initState() {
    super.initState();
    filteredFollowers = List.from(followers);
    _searchController.addListener(_filterFollowers);
  }

  void _filterFollowers() {
    final query = _searchController.text.toLowerCase();
    setState(() {
      if (query.isEmpty) {
        filteredFollowers = List.from(followers);
      } else {
        filteredFollowers = followers.where((user) {
          return user.username.toLowerCase().contains(query) ||
                 user.displayName.toLowerCase().contains(query);
        }).toList();
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text('Followers'),
            Text(
              '${followers.length} followers',
              style: TextStyle(fontSize: 14, color: Colors.grey[600]),
            ),
          ],
        ),
        backgroundColor: Colors.white,
        foregroundColor: Colors.black,
        elevation: 1,
      ),
      body: Column(
        children: [
          // Search bar
          Padding(
            padding: EdgeInsets.all(16),
            child: TextField(
              controller: _searchController,
              decoration: InputDecoration(
                hintText: 'Search followers...',
                prefixIcon: Icon(Icons.search),
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(8),
                  borderSide: BorderSide(color: Colors.grey[300]!),
                ),
                filled: true,
                fillColor: Colors.grey[100],
              ),
            ),
          ),
          // Categories (optional)
          Container(
            height: 40,
            child: ListView(
              scrollDirection: Axis.horizontal,
              padding: EdgeInsets.symmetric(horizontal: 16),
              children: [
                _buildCategoryChip('All', true),
                _buildCategoryChip('Following you back', false),
                _buildCategoryChip('Not following back', false),
              ],
            ),
          ),
          SizedBox(height: 8),
          // Followers list
          Expanded(
            child: RefreshIndicator(
              onRefresh: _refreshFollowers,
              child: ListView.builder(
                itemCount: filteredFollowers.length,
                itemBuilder: (context, index) {
                  return _buildFollowerItem(filteredFollowers[index]);
                },
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildCategoryChip(String label, bool isSelected) {
    return Padding(
      padding: EdgeInsets.only(right: 8),
      child: FilterChip(
        label: Text(label),
        selected: isSelected,
        onSelected: (selected) {
          // Implement filtering logic
          _filterByCategory(label);
        },
        selectedColor: Theme.of(context).primaryColor.withOpacity(0.2),
        checkmarkColor: Theme.of(context).primaryColor,
      ),
    );
  }

  Widget _buildFollowerItem(User user) {
    return ListTile(
      contentPadding: EdgeInsets.symmetric(horizontal: 16, vertical: 4),
      leading: GestureDetector(
        onTap: () => _navigateToProfile(user.username),
        child: CircleAvatar(
          radius: 24,
          backgroundImage: NetworkImage(user.avatar),
        ),
      ),
      title: GestureDetector(
        onTap: () => _navigateToProfile(user.username),
        child: Text(
          user.username,
          style: TextStyle(fontWeight: FontWeight.bold),
        ),
      ),
      subtitle: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(user.displayName),
          if (user.isFollowedBy && user.isFollowing)
            Text(
              'Follows you',
              style: TextStyle(fontSize: 12, color: Colors.grey),
            ),
        ],
      ),
      trailing: _buildFollowButton(user),
    );
  }

  Widget _buildFollowButton(User user) {
    if (user.username == 'current_user') {
      return SizedBox.shrink(); // Don't show button for current user
    }

    return SizedBox(
      width: 100,
      height: 32,
      child: OutlinedButton(
        onPressed: () => _toggleFollow(user),
        style: OutlinedButton.styleFrom(
          backgroundColor: user.isFollowing ? Colors.grey[200] : Colors.transparent,
          side: BorderSide(
            color: user.isFollowing ? Colors.grey : Theme.of(context).primaryColor,
          ),
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(8),
          ),
        ),
        child: Text(
          user.isFollowing ? 'Following' : 'Follow',
          style: TextStyle(
            color: user.isFollowing ? Colors.black : Theme.of(context).primaryColor,
            fontSize: 12,
            fontWeight: FontWeight.bold,
          ),
        ),
      ),
    );
  }

  void _filterByCategory(String category) {
    setState(() {
      switch (category) {
        case 'All':
          filteredFollowers = List.from(followers);
          break;
        case 'Following you back':
          filteredFollowers = followers.where((user) => user.isFollowing).toList();
          break;
        case 'Not following back':
          filteredFollowers = followers.where((user) => !user.isFollowing).toList();
          break;
      }
    });
  }

  Future<void> _refreshFollowers() async {
    setState(() {
      isLoading = true;
    });

    // Simulate API call
    await Future.delayed(Duration(seconds: 2));

    // In real app, fetch updated followers list from server
    setState(() {
      isLoading = false;
    });
  }

  void _toggleFollow(User user) {
    setState(() {
      user.isFollowing = !user.isFollowing;
    });

    // Show feedback
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text(
          user.isFollowing
            ? 'You are now following ${user.username}'
            : 'You unfollowed ${user.username}',
        ),
        duration: Duration(seconds: 2),
      ),
    );
  }

  void _navigateToProfile(String username) {
    Navigator.pushNamed(
      context,
      '/user-profile',
      arguments: username,
    );
  }

  @override
  void dispose() {
    _searchController.dispose();
    super.dispose();
  }
}

class User {
  final String id;
  final String username;
  final String displayName;
  final String avatar;
  bool isFollowing;
  final bool isFollowedBy;

  User({
    required this.id,
    required this.username,
    required this.displayName,
    required this.avatar,
    required this.isFollowing,
    required this.isFollowedBy,
  });
}

File: snippets/screens/swiftui/followers_screen/3/followers_screen.swift
import SwiftUI

struct FollowersScreen: View {
    let username: String
    @StateObject private var viewModel = FollowersViewModel()
    @State private var searchText = ""

    var body: some View {
        NavigationView {
            VStack(spacing: 0) {
                // Search bar
                SearchBar(text: $searchText)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 12)

                // Category filters
                CategoryFilterBar(
                    selectedCategory: $viewModel.selectedCategory,
                    categories: viewModel.categories
                )

                // Followers list
                List {
                    ForEach(viewModel.filteredFollowers) { follower in
                        FollowerRow(
                            user: follower,
                            onToggleFollow: {
                                viewModel.toggleFollow(for: follower.id)
                            },
                            onUserTap: {
                                // Navigate to user profile
                            }
                        )
                        .listRowInsets(EdgeInsets(top: 8, leading: 16, bottom: 8, trailing: 16))
                    }
                }
                .listStyle(.plain)
                .refreshable {
                    await viewModel.refreshFollowers()
                }
            }
            .navigationTitle(viewModel.navigationTitle)
            .navigationBarTitleDisplayMode(.inline)
        }
        .onAppear {
            viewModel.loadFollowers(for: username)
        }
        .onChange(of: searchText) { text in
            viewModel.filterFollowers(searchText: text)
        }
    }
}

struct SearchBar: View {
    @Binding var text: String

    var body: some View {
        HStack {
            Image(systemName: "magnifyingglass")
                .foregroundColor(.secondary)

            TextField("Search followers...", text: $text)
                .textFieldStyle(.plain)
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
        .background(Color.gray.opacity(0.1))
        .cornerRadius(10)
    }
}

struct CategoryFilterBar: View {
    @Binding var selectedCategory: String
    let categories: [String]

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 12) {
                ForEach(categories, id: \.self) { category in
                    FilterChip(
                        title: category,
                        isSelected: selectedCategory == category,
                        onTap: {
                            selectedCategory = category
                        }
                    )
                }
            }
            .padding(.horizontal, 16)
        }
        .padding(.vertical, 8)
    }
}

struct FilterChip: View {
    let title: String
    let isSelected: Bool
    let onTap: () -> Void

    var body: some View {
        Button(action: onTap) {
            Text(title)
                .font(.subheadline)
                .padding(.horizontal, 16)
                .padding(.vertical, 8)
                .background(isSelected ? Color.blue.opacity(0.1) : Color.gray.opacity(0.1))
                .foregroundColor(isSelected ? .blue : .primary)
                .overlay(
                    RoundedRectangle(cornerRadius: 20)
                        .stroke(isSelected ? Color.blue : Color.clear, lineWidth: 1)
                )
                .cornerRadius(20)
        }
    }
}

struct FollowerRow: View {
    let user: UserModel
    let onToggleFollow: () -> Void
    let onUserTap: () -> Void

    var body: some View {
        HStack(spacing: 12) {
            Button(action: onUserTap) {
                AsyncImage(url: URL(string: user.avatar)) { image in
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fill)
                } placeholder: {
                    Circle()
                        .fill(Color.gray.opacity(0.3))
                }
                .frame(width: 48, height: 48)
                .clipShape(Circle())
            }

            VStack(alignment: .leading, spacing: 2) {
                Button(action: onUserTap) {
                    Text(user.username)
                        .font(.headline)
                        .fontWeight(.semibold)
                        .foregroundColor(.primary)
                }

                Text(user.displayName)
                    .font(.subheadline)
                    .foregroundColor(.secondary)

                if user.isFollowedBy && user.isFollowing {
                    Text("Follows you")
                        .font(.caption)
                        .foregroundColor(.secondary)
                }
            }

            Spacer()

            if user.username != "current_user" {
                Button(user.isFollowing ? "Following" : "Follow") {
                    onToggleFollow()
                }
                .font(.subheadline)
                .fontWeight(.semibold)
                .foregroundColor(user.isFollowing ? .primary : .white)
                .padding(.horizontal, 20)
                .padding(.vertical, 8)
                .background(user.isFollowing ? Color.gray.opacity(0.2) : Color.blue)
                .cornerRadius(8)
            }
        }
    }
}

@MainActor
class FollowersViewModel: ObservableObject {
    @Published var followers: [UserModel] = []
    @Published var filteredFollowers: [UserModel] = []
    @Published var selectedCategory = "All"
    @Published var navigationTitle = "Followers"

    let categories = ["All", "Following you back", "Not following back"]

    private var allFollowers: [UserModel] = []
    private var currentSearchText = ""

    func loadFollowers(for username: String) {
        // Sample followers data
        allFollowers = [
            UserModel(
                id: "1",
                username: "alice_wonder",
                displayName: "Alice Wonder",
                avatar: "https://example.com/avatar3.jpg",
                isFollowing: true,
                isFollowedBy: true
            ),
            UserModel(
                id: "2",
                username: "bob_builder",
                displayName: "Bob Builder",
                avatar: "https://example.com/avatar4.jpg",
                isFollowing: false,
                isFollowedBy: true
            ),
            UserModel(
                id: "3",
                username: "charlie_brown",
                displayName: "Charlie Brown",
                avatar: "https://example.com/avatar5.jpg",
                isFollowing: true,
                isFollowedBy: true
            ),
            UserModel(
                id: "4",
                username: "diana_prince",
                displayName: "Diana Prince",
                avatar: "https://example.com/avatar6.jpg",
                isFollowing: false,
                isFollowedBy: true
            ),
            UserModel(
                id: "5",
                username: "edward_stark",
                displayName: "Edward Stark",
                avatar: "https://example.com/avatar7.jpg",
                isFollowing: true,
                isFollowedBy: true
            )
        ]

        followers = allFollowers
        navigationTitle = "\(followers.count) followers"
        applyFilters()
    }

    func filterFollowers(searchText: String) {
        currentSearchText = searchText
        applyFilters()
    }

    func applyFilters() {
        var filtered = allFollowers

        // Apply category filter
        switch selectedCategory {
        case "Following you back":
            filtered = filtered.filter { $0.isFollowing }
        case "Not following back":
            filtered = filtered.filter { !$0.isFollowing }
        default:
            break
        }

        // Apply search filter
        if !currentSearchText.isEmpty {
            filtered = filtered.filter { user in
                user.username.localizedCaseInsensitiveContains(currentSearchText) ||
                user.displayName.localizedCaseInsensitiveContains(currentSearchText)
            }
        }

        filteredFollowers = filtered
    }

    func toggleFollow(for userId: String) {
        if let index = allFollowers.firstIndex(where: { $0.id == userId }) {
            allFollowers[index].isFollowing.toggle()
            applyFilters()
        }
    }

    func refreshFollowers() async {
        // Simulate network request
        try? await Task.sleep(nanoseconds: 1_000_000_000)
        // Refresh data if needed
    }
}

struct UserModel: Identifiable {
    let id: String
    let username: String
    let displayName: String
    let avatar: String
    var isFollowing: Bool
    let isFollowedBy: Bool
}

#Preview {
    FollowersScreen(username: "john_doe")
}